implementations/rust/ockam/ockam_api/src/nodes/service/http.rs

```diff
@@ -0,0 +1,119 @@
+use std::convert::Infallible;
+use std::net::SocketAddr;
+use std::sync::Arc;
+
+use http_body_util::combinators::BoxBody;
+use http_body_util::{BodyExt, Empty, Full};
+use hyper::body::Bytes;
+use hyper::server::conn::http1;
+use hyper::service::service_fn;
+use hyper::{Method, Request, Response};
+use hyper_util::rt::TokioIo;
+use ockam_core::AsyncTryClone;
+use ockam_node::Context;
+use serde::Serialize;
+use tokio::net::TcpListener;
+use tokio::task::JoinHandle;
+
+use crate::nodes::models::node::NodeResources;
+use crate::nodes::BackgroundNodeClient;
+use crate::{HttpError, Result};
+
+pub struct HttpServer {
+    ctx: Arc<Context>,
+    node_client: Arc<BackgroundNodeClient>,
```

Comment:
Maybe I misunderstood something, but since the server is already running inside a node, it should need a client to access the node data. The `get_node_resources` function is directly available on the `NodeManager`.

Reply:
I started with that approach, but I got some issues. The `NodeManager` needs to keep the `JoinHandle` of the future task that runs the `HttpServer` to shut down gracefully. If I want to keep a `NodeManager` instance inside the `HttpServer`, then I need to create first the `NodeManager`, arc it, start the `HttpServer` and mutate the original `NodeManager`. That felt too complex, and I opted to decouple them using the `BackgroundNodeClient` as the intermediary.

Reply:
Would it be the same if you used the `NodeManagerWorker` and sent a local message instead?

Reply:
Or maybe initialize the HttpServer with references to the `registry`, the `transport` and the `cli_state` from the `NodeManager` because that's all that it is required to return the nodes resources?

Reply:
> Would it be the same if you used the NodeManagerWorker and sent a local message instead?

That has the same problem as the `NodeManager` (you need the node manager to be initialized).

> Or maybe initialize the HttpServer with references to the registry, the transport and the cli_state from the NodeManager because that's all that is required to return the nodes resources?

I think that could be assuming too much from the HttpServer point of view. What if we also want to retrieve some data outside of those objects' scope (e.g. some project data)? The cost of keeping a BackgroundNodeClient copy is minimal and it doesn't limit us.

Reply:
It feels really wrong to me to fully step outside of the Rust process, go through the whole transport layer, to access some information that is already there.

> What if we also want to retrieve some data outside of those objects' scope (e.g. some project data)? The cost of keeping a BackgroundNodeClient copy is minimal and it doesn't limit us.

I think that in that case we can (and we should) start making high-level APIs for the `NodeManager` instead of a monolithic set of routes redirecting to a "God" object.

Reply:
Ok, let me try to move some things around to include the NodeManager inside the HttpServer.